### ISSUE TITLE
[Bug] Unable to use OptunaSweeperConf using the Python API

### ISSUE THREAD

**Issue by mayeroa** [facebookresearch/hydra#1830]
# 🐛 Bug
## Description

Hi,

First of all, many thanks for that wonderful tool you've made. It is, for sure, extremely helpful to have it when production environment is necessary.
I have met an error/bug when I wanted to use the **OptunaSweeperConf** using the Python API (instead of using the traditional YAML files). 

## Checklist
- [x] I checked on the latest version of Hydra
- [x] I created a minimal repro (See [this](https://stackoverflow.com/help/minimal-reproducible-example) for tips).

## To reproduce

** Minimal Code/Config snippet to reproduce **

```python
# Standard libraries
from dataclasses import dataclass

# Third-party libraries
import hydra
from hydra.conf import HydraConf
from hydra.core.config_store import ConfigStore
from hydra_plugins.hydra_optuna_sweeper.config import OptunaSweeperConf
from omegaconf import DictConfig


@dataclass
class ExampleConfig():
    hydra: HydraConf = HydraConf(sweeper=OptunaSweeperConf())


ConfigStore.instance().store(name='config', node=ExampleConfig)


@hydra.main(config_path=None, config_name='config')
def main(config: DictConfig) -> None:
    """Main function."""
    print(config)


if __name__ == '__main__':
    main()

```

And the stack trace with the returned error:

** Stack trace/error message **
```
In 'config': Validation error while composing config:
Merge error: OptunaSweeperConf is not a subclass of BasicSweeperConf. value: {'_target_': 'hydra_plugins.hydra_optuna_sweeper.optuna_sweeper.OptunaSweeper', 'defaults': [{'sampler': 'tpe'}], 'sampler': '???', 'direction': <Direction.minimize: 1>, 'storage': None, 'study_name': None, 'n_trials': 20, 'n_jobs': 2, 'search_space': {}}
    full_key: 
    object_type=dict

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

## Expected Behavior

I would like to override the hydra config group, directly into the structure config. To do so, I import the **HydraConf** from hydra and install the **hydra-optuna-sweeper** plugin.
The error seems to indicate that the **OptunaSweeperConf** must inherit from the **BasicSweeperConf** structured config (current implementation: https://github.com/facebookresearch/hydra/blob/v1.1.1/plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/config.py#L133).
The idea is to be able to override hydra config group directly from a python script. Does that make sense?

## System information
- **Hydra Version** :  1.1.1
- **Hydra Optuna Sweeper Version** :  1.1.1
- **Python version** : 3.8.11
- **Virtual environment type and version** : MiniConda environment
- **Operating system** : LinuxMint

**Comment by jieru-hu** [comment]
hi @mayeroa  pls try using the defaults list instead. something like the following should work
```
@dataclass
class ExampleConfig():
    defaults: List[Any] = field(
        default_factory=lambda: [
            {"override hydra/sweeper": "optuna"},
        ]
    )

```

Run the application, you should see optuna's config
```
$python my_app.py --cfg hydra -p hydra.sweeper
# @package hydra.sweeper
sampler:
  _target_: optuna.samplers.TPESampler
  seed: null
  consider_prior: true
  prior_weight: 1.0
  consider_magic_clip: true
  consider_endpoints: false
  n_startup_trials: 10
  n_ei_candidates: 24
  multivariate: false
  warn_independent_sampling: true
_target_: hydra_plugins.hydra_optuna_sweeper.optuna_sweeper.OptunaSweeper
direction: minimize
storage: null
study_name: null
n_trials: 20
n_jobs: 2
search_space: {}
```


Also the `cfg` object you get via `hydra.main` does not contain Hydra's config, to access the configs, try something like the following
```
@hydra.main(config_path=None, config_name='config')
def main(config: DictConfig) -> None:
    """Main function."""
    print(HydraConfig.get().sweeper)
```

**Comment by mayeroa** [comment]
Thank you for your answer @jieru-hu, it helps.
And if I want to override some parameters of the `OptunaSweeper` directly into the `ExampleConfig`, how should I proceed?

For example : 

```python
@dataclass
class ExampleConfig():
    defaults: List[Any] = field(
        default_factory=lambda: [
            {"override hydra/sweeper": "optuna"},
        ]
    )
# How to override n_trials?
# How to define search space?
```

**Comment by jieru-hu** [comment]
hi @mayeroa 

> And if I want to override some parameters of the OptunaSweeper directly into the ExampleConfig, how should I proceed?

We do have in detailed documentation on how you'd configure a plugin, pls check here https://hydra.cc/docs/patterns/configuring_plugins

> I think, it would be of great value to support such sweeper instantiation directly in python code. What do you think?

Essentially you want to override Hydra configuration here right? You can configure Hydra just like how you'd configure your application configuration. For example, you can write it in your config files, or override it from the commandline. 

```
# in your primary config
hydra:
  run:
     dir:
          ./my_dir

```
or from the commandline
```
hydra.run.dir=./my_dir
```

Hydra takes care of the config composition for you so you do not need to manage them yourself. I'd suggest go through our basic tutorial and the configure Hydra section of the doc if you have not already
https://hydra.cc/docs/next/tutorials/intro/
https://hydra.cc/docs/next/configure_hydra/intro

**Comment by mayeroa** [comment]
Hi @jieru-hu,
Sorry for the explanation, I am very unclear.
I am perfectly aware of how to configure a sweeper using yaml files or directly through the command the line.
My question was about configuring a Sweeper directly using the Python API (structured config)?
But your answers suggest that it is not possible or not proposed by hydra.

To put it in some context, I'm using Hydra to run Deep Learning trainings using PyTorch-Lightning like that:

```python
@dataclass
class ExampleConfig:

    datamodule = DataModuleConf(
        train_dataset=RandomDatasetConf(),
        val_dataset=RandomDatasetConf(),
        test_dataset=RandomDatasetConf(),
        num_workers=0,
        batch_size=8
    )

    module = ModuleConf(
        model=RandomModelConf(),
        loss=CrossEntropyLossConf(),
        optimizer=SGDConf(lr=0.1),
        scheduler=StepLRConf(step_size=1),
    )

    trainer = TrainerConf(
        callbacks=[LearningRateMonitorConf()],
        max_epochs=20
    )


ConfigStore.instance().store(name='config', node=ExampleConfig)


@hydra.main(config_path=None, config_name='config')
def main(config: DLCookerConf) -> None:
    """Main function."""
    # Set seed
    seed_everything(seed=config.training.seed, workers=True)

    # Instantiate datamodule
    datamodule = instantiate(config=config.datamodule)

    # Instantiate module
    module = instantiate(config=config.module)

    # Instantiate trainer
    trainer = instantiate(config=config.trainer)

    # Run training & test
    trainer.fit(module, datamodule=datamodule)
    trainer.test(module, datamodule=datamodule)


if __name__ == '__main__':
    main()
```

I really like to use the Python API because it offers the benefits of using an IDE (like VSCode) to easily manipulate structured configurations with auto-completion.
And I would like to be happy to override **HydraConf** directly in that file (having a single file to centralize all the config used for training). I managed to override the run dir like so, with adding the **hydra** config group: 

```python
    hydra: HydraConf = HydraConf(run=RunDir('./test'))
```

But if I want to do the same thing with a sweeper (like Optuna), it fails (see the error at the beginning of the thread).
I understand that Hydra relies essentially on manipulating and composing YAML files but I also see an interest of manipulating the structured config (the schemas) directly through a Python script. 
Maybe I am wrong 😃 . 

PS: there was an attempt, in https://github.com/pytorch/hydra-torch/blob/main/examples/mnist_00.py to use directly structured configs through a script. And I think we can go a step further with the example above. What do you think?

**Comment by Jasha10** [comment]
> ```
> # How to override n_trials?
> # How to define search space?
> ```

I have tried the following, which yields a surprising error:
```python
from dataclasses import dataclass, field

import hydra
from hydra.conf import HydraConf
from hydra.core.config_store import ConfigStore
from hydra_plugins.hydra_optuna_sweeper.config import OptunaSweeperConf
from omegaconf import DictConfig
from typing import List
from typing import Any

# Third-party libraries


@dataclass
class ExampleConfig:
    defaults: List[Any] = field(
        default_factory=lambda: [
            {"override hydra/sweeper": "optuna"},
        ]
    )
    hydra: HydraConf = HydraConf(sweeper=OptunaSweeperConf(n_trials=100))

ConfigStore.instance().store(name="config", node=ExampleConfig)


@hydra.main(config_path=None, config_name="config")
def main(config: DictConfig) -> None:
    """Main function."""
    print(config)


if __name__ == "__main__":
    main()
```
Here's the output:
```text
$ python my_app.py
In 'config': ValidationError raised while composing config:
Merge error: OptunaSweeperConf is not a subclass of OptunaSweeperConf. value: {'_target_': 'hydra_plugins.hydra_optuna_sweeper.optuna_sweeper.OptunaSweeper', 'defaults': [{'sampler': 'tpe'}], 'sampler': '???', 'direction': <Direction.minimize: 1>, 'storage': None, 'study_name': None, 'n_trials': 100, 'n_jobs': 2, 'search_space': {}}
    full_key:
    object_type=dict

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```
Note in particular that the error message says "OptunaSweeperConf is not a subclass of OptunaSweeperConf." This smells fishy to me... I think there might be a bug here.

**Comment by omry** [comment]
This is probably related to how plugins are loaded discovered and loaded.
There are two different OptunaSweeperConf classes loaded.

**Comment by mayeroa** [comment]
> This is probably related to how plugins are loaded discovered and loaded. There are two different OptunaSweeperConf classes loaded.

Do you think I can have a look on it or does it seem like a complex fix to implement?

**Comment by jieru-hu** [comment]
sorry @mayeroa for getting back to you late, thanks for further explaining your use case.

like @omry suspected earlier, it does seem to be related to how the two classes are imported, in particular, the application fails at merging the configs [here](https://github.com/omry/omegaconf/blob/e3ea9a117f49595c331b399dfe45d0423c7e45e9/omegaconf/dictconfig.py#L234), although the two Conf classes have the same type, the `issubclass` return False.

I tried to move the [`ConfigStore` call](https://github.com/facebookresearch/hydra/blob/main/plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/config.py#L164-L169) to the `__init__.py` like the following

```
a/plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/__init__.py
+++ b/plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/__init__.py
@@ -1,3 +1,14 @@
 # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
 
 __version__ = "1.2.0dev1"
+
+from hydra_plugins.hydra_optuna_sweeper.config import OptunaSweeperConf
+
+from hydra.core.config_store import ConfigStore
+
+ConfigStore.instance().store(
+    group="hydra/sweeper",
+    name="optuna",
+    node=OptunaSweeperConf,
+    provider="optuna_sweeper",
+)
``` 
which seems to resolve the issue, and the config override worked
running @Jasha10's example above
```
$ python /Users/jieru/workspace/hydra-fork/hydra/examples/tutorials/structured_configs/1_minimal/my_app.py --cfg hydra -p hydra.sweeper.n_trials
100
```

I'm not sure if this is the best way to fix this though, @omry thoughts?

**Comment by omry** [comment]
Worth trying to figure out the root cause.

**Comment by mayeroa** [comment]
Hello :)
Do you have any update on that topic? Or need some help to further investigate on that?

**Comment by jieru-hu** [comment]
hi @mayeroa thanks for the ping, we will get to this for the next release. in the meanwhile, feel free to further investigate and maybe even suggest a fix :)

**Comment by jieru-hu** [comment]
Ok - I took another look at this and here is what is happening:
the tl;dr is the plugins are being imported twice and as a result OmegaConf fails the validation [here](https://github.com/omry/omegaconf/blob/master/omegaconf/dictconfig.py#L230)

1. OmegaConf fails [here](https://github.com/omry/omegaconf/blob/master/omegaconf/dictconfig.py#L207) - complaining  `OptunaSweeperConf is not a subclass of OptunaSweeperConf`. In debug mode, comparing the dest_obj_type and src_obj_type, i can see that although they are the *same* class, they have different ids (meaning they are different types due to the double imports)
2. This seems to be only happening to plugins, since they are being imported twice: 1) first time in the user application 2) second time [here](https://github.com/facebookresearch/hydra/blob/dd2783dacc719fb88935be729cdb96d21073f0ac/hydra/core/plugins.py#L170-L172)
3. This is probably not an optuna only issue but affects all plugins.

I also found a so question that describes a similar issue [here](https://stackoverflow.com/questions/11461356/issubclass-returns-false-on-the-same-class-imported-from-different-paths)

**Comment by Jasha10** [comment]
Nice detective work!!

Here is a minimal reproducer for the issue:
```python
# min_repo1.py
import importlib
import pkgutil

from hydra_plugins.hydra_optuna_sweeper.config import OptunaSweeperConf

mdl = importlib.import_module("hydra_plugins")
for importer, modname, ispkg in pkgutil.walk_packages(
    path=mdl.__path__, prefix=mdl.__name__ + ".", onerror=lambda x: None
):
    m = importer.find_module(modname)
    loaded_mod = m.load_module(modname)
    if loaded_mod.__name__ == "hydra_plugins.hydra_optuna_sweeper.config":
        assert loaded_mod.OptunaSweeperConf is OptunaSweeperConf
        print("assertion passed!")
```
```text
$ python min_repo1.py
Traceback (most recent call last):
  File "min_repo1.py", line 14, in <module>
    assert loaded_mod.OptunaSweeperConf is OptunaSweeperConf, "comparison failed"
AssertionError: comparison failed
```

I have an idea for a diff to solve the issue.
```diff
$ diff min_repo1.py min_repo2.py
1c1
< # min_repo1.py
---
> # min_repo2.py
11,12c11
<     m = importer.find_module(modname)
<     loaded_mod = m.load_module(modname)
---
>     loaded_mod = importlib.import_module(modname)
```
```text
$ python min_repo2.py
assertion passed!
```
By doing the import with `importlib.import_module` instead of `pkgutil`, we get the same module as is loaded by the `import` statement at the top of the file.

**Comment by mayeroa** [comment]
Awesome investigation!
That'a a bit weird by the way because I thought `pkgutil` was using `importlib` under the hood. Obviously not.
So changing the way we import module should be enough to fix the issue? Do you want me to send a PR or does it need further investigation (and another solution)?

**Comment by jieru-hu** [comment]
Thanks @Jasha10 . It's interesting to see all these tiny (and undocumented 🙃 ) subtleties between the standard libs.

Another part of the root cause is we call `ConfigStore.store` twice: 1) first time when users import the plugin conf `dataclasses`. 2) when scanning all plugins we try to import module under the hydra_plugin namespace except for the ones starts with `__` and `_`, which explains why in https://github.com/facebookresearch/hydra/issues/1830#issuecomment-933914988 when I put the ConfigStore code in `__init__.py`, everything works fine. (`__init__.py` is not imported again in `Plugins`)

updating `loaded_mod = importlib.import_module(modname)` looks good to me.

**Comment by Jasha10** [comment]
> `__init__.py` is not imported again in Plugins

Single underscores are ignored but double underscores are not ignored.
```python
if module_name.startswith("_") and not module_name.startswith("__"):
```
([ref](https://github.com/facebookresearch/hydra/blob/dd2783dacc719fb88935be729cdb96d21073f0ac/hydra/core/plugins.py#L167))

So `__init__.py` should still be imported.
I am confused xD

**Comment by jieru-hu** [comment]
@Jasha10 you are right! 🤣  i updated the previous comment.

**Comment by Jasha10** [comment]
> So changing the way we import module should be enough to fix the issue?

I think changing the way we import the module is necessary, and hopefully it will be sufficient too.

> Do you want me to send a PR or does it need further investigation (and another solution)?

Thanks for the offer :)
I've just opened PR #2019.
@mayeroa, would you mind testing it out to see if that diff is sufficient to solve the problem that was coming up in your use-case?

**Comment by mayeroa** [comment]
Hi @Jasha10,

I have made some tests using your PR with the following example:

```python
# Standard libraries
from dataclasses import dataclass, field
from typing import Any, List

# Third-party libraries
import hydra
from hydra.conf import HydraConf
from hydra.core.config_store import ConfigStore
from hydra_plugins.hydra_optuna_sweeper.config import OptunaSweeperConf
from omegaconf import DictConfig



@dataclass
class ExampleConfig:
    defaults: List[Any] = field(
        default_factory=lambda: [
            {"override hydra/sweeper": "optuna"},
        ]
    )
    hydra: HydraConf = HydraConf(sweeper=OptunaSweeperConf(n_trials=10))

ConfigStore.instance().store(name="config", node=ExampleConfig)


@hydra.main(config_path=None, config_name="config")
def main(config: DictConfig) -> None:
    """Main function."""
    return 2.0


if __name__ == "__main__":
    main()

``` 

Running with following command `python test.py --multirun`, I get this output:

```shell
[I 2022-02-08 09:24:18,271] A new study created in memory with name: no-name-9010883b-3493-4a59-921c-ee32feb7651a
[2022-02-08 09:24:18,272][HYDRA] Study name: no-name-9010883b-3493-4a59-921c-ee32feb7651a
[2022-02-08 09:24:18,272][HYDRA] Storage: None
[2022-02-08 09:24:18,272][HYDRA] Sampler: TPESampler
[2022-02-08 09:24:18,272][HYDRA] Directions: ['minimize']
[2022-02-08 09:24:18,274][HYDRA] Launching 2 jobs locally
[2022-02-08 09:24:18,274][HYDRA]        #0 : 
[2022-02-08 09:24:18,396][HYDRA]        #1 : 
[2022-02-08 09:24:18,560][HYDRA] Launching 2 jobs locally
[2022-02-08 09:24:18,560][HYDRA]        #2 : 
[2022-02-08 09:24:18,664][HYDRA]        #3 : 
[2022-02-08 09:24:18,769][HYDRA] Launching 2 jobs locally
[2022-02-08 09:24:18,769][HYDRA]        #4 : 
[2022-02-08 09:24:18,875][HYDRA]        #5 : 
[2022-02-08 09:24:18,980][HYDRA] Launching 2 jobs locally
[2022-02-08 09:24:18,980][HYDRA]        #6 : 
[2022-02-08 09:24:19,086][HYDRA]        #7 : 
[2022-02-08 09:24:19,209][HYDRA] Launching 2 jobs locally
[2022-02-08 09:24:19,209][HYDRA]        #8 : 
[2022-02-08 09:24:19,317][HYDRA]        #9 : 
[2022-02-08 09:24:19,422][HYDRA] Best parameters: {}
[2022-02-08 09:24:19,422][HYDRA] Best value: 2.0
```

Which seems to be fine for me. :)

**Comment by Jasha10** [comment]
Great!

**Comment by Jasha10** [comment]
We are deferring work on this due to unforeseen complexity (see comments on PR #2019).

**Comment by mayeroa** [comment]
#2697

**Comment by omry** [comment]
Thanks for the detailed report, and sorry for the long delay.

In Hydra 1.4, the Optuna Sweeper can be configured entirely with structured
configs while keeping everything in Python.

The sweeper is a config-group choice, so the customized structured config
should be registered under `hydra/sweeper` and selected through the Defaults
List. Assigning `OptunaSweeperConf` directly to `HydraConf.sweeper` does not
replace the default `hydra/sweeper=basic` choice, which is why Hydra tries to
merge it into `BasicSweeperConf`.

```python
from dataclasses import dataclass, field
from typing import Any, List

import hydra
from hydra.core.config_store import ConfigStore
from hydra.core.hydra_config import HydraConfig
from hydra_plugins.hydra_optuna_sweeper.config import (
    OptunaSweeperConf,
    TPESamplerConfig,
)
from omegaconf import DictConfig


cs = ConfigStore.instance()

cs.store(
    group="hydra/sweeper",
    name="custom_optuna",
    node=OptunaSweeperConf(
        defaults=[],
        sampler=TPESamplerConfig(seed=123),
        n_trials=20,
        n_jobs=1,
        study_name="sphere",
        direction="minimize",
        params={
            "x": "range(-5.5, 5.5, step=0.5)",
            "y": "choice(-5, 0, 5)",
        },
    ),
)


@dataclass
class ExampleConfig:
    defaults: List[Any] = field(
        default_factory=lambda: [
            {"override hydra/sweeper": "custom_optuna"},
        ]
    )

    x: float = 1.0
    y: float = 1.0


cs.store(name="config", node=ExampleConfig)


@hydra.main(version_base=None, config_path=None, config_name="config")
def main(cfg: DictConfig) -> float:
    print(HydraConfig.get().sweeper)
    return cfg.x**2 + cfg.y**2


if __name__ == "__main__":
    main()
```

This configures `n_trials`, the search space, and the sampler through Python
structured configs. Run it with `python app.py --multirun`. The `cfg` argument
contains the application configuration, while Hydra's composed configuration is
available through `HydraConfig.get()`.

I verified this pattern with Hydra 1.4 and am closing the issue as completed.
If you encounter another problem with this approach, please open a new focused
issue with a minimal reproduction.
